### PR TITLE
Fixing Issue #4480, ReleaseHandle() getting called more than once due to probable race condition

### PR DESCRIPTION
**Pull request by rajansingh10**
Fix for issue #4480.
Looks like this is happening due to probable race condition which leads to ReleaseHandle() getting called more than once. ReleaseHandle was not setting holding object/pointer to null/Intptr. Setting it explicitly now. 
@roncain can you please validate this fix with your test which was crashing earlier.

### CODE REVIEW

**Comment by rajansingh10** [comment]
cc: @stephentoub, @bartonjs, @davidsh, @CIPop, @SidharthNabar, @pgavlin , @vijaykota

**Comment by stephentoub** [comment]
ReleaseHandle is protected by an interlocked operation in the runtime and will never be called more than once.  Setting handle to IntPtr.Zero in ReleaseHandle will cause IsInvalid to return true after the handle has been closed, but it won't change anything about ReleaseHandle being invoked twice, which it won't be.

File: src/Common/src/Interop/Unix/libssl/SecuritySafeHandles.cs [hunk 1]
```diff
@@ -304,6 +308,7 @@ protected override bool ReleaseHandle()
         {
             _channelBinding.DangerousRelease();
             _channelBinding.Dispose();
+            _channelBinding = null;
```
Comment (vijaykota) [inline src/Common/src/Interop/Unix/libssl/SecuritySafeHandles.cs]: IsInvalid should now check for null

**Comment by vijaykota** [comment]
@stephentoub , since there is no repro, would the better fix then be to _sslContext.AddRef() in SafeDeleteContext and _sslContext.Release in SafeDeleteContext.ReleaseHandle?

@roncain, would it be possible for you to try that out?

File: src/Common/src/Interop/Unix/libssl/SecuritySafeHandles.cs (hunk 1 above)
Comment (rajansingh10) [inline src/Common/src/Interop/Unix/libssl/SecuritySafeHandles.cs]: ok

**Comment by davidsh** [comment]
>but it won't change anything about ReleaseHandle being invoked twice, which it won't be.

Totally agree with this part.  The SafeHandle base class is pretty robust and the derived classes' `ReleaseHandle` method will not be called twice.

**Comment by stephentoub** [comment]
>  since there is no repro

It sounded like there was a repro, and that @roncain is able to repro it at will, it's just a complicated one.

> would the better fix then be to

I've not investigated the problem, but the current "fix" doesn't actually appear to be one, per my previous comments.  That said, your suggestion does sound plausible.  In general any time there’s a hierarchical relationship between SafeHandles, where a parent SafeHandle stores a reference to a child SafeHandle and where the parent expects the child to survive at least as long as the parent, the parent SafeHandle should DangerousAddRef the child when the child is stored into it and should DangerousRelease the child during the parent’s ReleaseHandle.  If that's the relationship here and if such add/release ref'ing isn't being done, then yes, it should be.

**Comment by davidsh** [comment]
>any time there’s a hierarchical relationship between SafeHandles, where a parent SafeHandle stores a reference to a child SafeHandle and where the parent expects the child to survive at least as long as the parent, the parent SafeHandle should DangerousAddRef the child when the child is stored into it and should DangerousRelease the child during the parent’s ReleaseHandle.

I use this technique in WinHttpHandler to manage parent-child relationships among the WinHTTP handles (session -> connect -> request)

**Comment by davidsh** [comment]
Although the way I do it in WinHttpHandler is that the child SafeHandle stores a reference to the parent.  And in doing so, DangerousAddRef's the parent as part of storing the reference to it in the child.

See: https://github.com/dotnet/corefx/blob/master/src/Common/src/Interop/Windows/winhttp/Interop.SafeWinHttpHandle.cs

**Comment by rajansingh10** [comment]
closing this. Will send another PR with right fix.

**Pull request closed without merging**